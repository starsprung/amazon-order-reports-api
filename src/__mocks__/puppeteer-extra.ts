import { Browser, CDPSession, Page, Response, Target } from 'puppeteer';

const response = ({
  ok: jest.fn()
} as unknown) as Response;

const cdpSession = ({
  send: jest.fn()
} as unknown) as CDPSession;

const target = ({ createCDPSession: jest.fn() } as unknown) as Target;

const page = ({
  $: jest.fn(),
  $x: jest.fn(),
  $$eval: jest.fn(),
  click: jest.fn(),
  goto: jest.fn(),
  select: jest.fn(),
  target: jest.fn(),
  type: jest.fn(),
  url: jest.fn(),
  waitForNavigation: jest.fn(),
  waitForResponse: jest.fn()
} as unknown) as Page;

const browser = ({
  close: jest.fn(),
  newPage: jest.fn()
} as unknown) as Browser;

export const mocks = {
  browser,
  cdpSession,
  page,
  response,
  target
};

export default {
  launch: jest.fn(),
  use: jest.fn()
};
