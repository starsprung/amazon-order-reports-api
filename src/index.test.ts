import appRootPath from 'app-root-path';
import { createReadStream, Dirent, FSWatcher, PathLike, ReadStream, watch } from 'fs';
import { mkdtemp, readdir, unlink } from 'fs/promises';
import mockdate from 'mockdate';
import { tmpdir } from 'os';
import { ElementHandle, Response } from 'puppeteer';
import puppeteer from 'puppeteer-extra';
import { Readable } from 'stream';
import { mocked } from 'ts-jest/utils';
import AmazonScraper, { OrderItem } from './index';
import { mocks } from './__mocks__/puppeteer-extra';
import { v4 as uuidv4 } from 'uuid';

jest.mock('fs');
jest.mock('fs/promises');
jest.mock('os');
jest.mock('uuid');

describe('AmazonScraper', () => {
  beforeEach(() => {
    mocked(createReadStream).mockReturnValue((Readable.from([]) as unknown) as ReadStream);
    mocked(mkdtemp).mockResolvedValue('/tmp/amzscrKudNUt');
    mocked(mocks.browser.newPage).mockResolvedValue(mocks.page);
    mocked(mocks.page.$).mockResolvedValue(({} as unknown) as ElementHandle<Element>);
    mocked(mocks.page.goto).mockResolvedValue(mocks.response);
    mocked(mocks.page.target).mockReturnValue(mocks.target);
    mocked(mocks.page.url).mockReturnValue('');
    mocked(mocks.page.waitForResponse).mockImplementation(async (fn) => {
      (fn as (response: Response) => boolean)(({
        url: () => 'https://www.amazon.com/b2b/reports/download'
      } as unknown) as Response);
      return {} as Response;
    });
    mocked(mocks.response.ok).mockReturnValue(true);
    mocked(mocks.target.createCDPSession).mockResolvedValue(mocks.cdpSession);
    mocked(puppeteer.launch).mockResolvedValue(mocks.browser);
    mocked(readdir).mockResolvedValue([('01-Dec-2020_to_31-Dec-2020.csv' as unknown) as Dirent]);
    mocked(tmpdir).mockReturnValue('/tmp/');
    mocked(uuidv4).mockReturnValue('5fb041e4-ad7a-41d4-879f-d1ec1919201a');
    mocked(watch).mockImplementation((_filename: PathLike, fn) => {
      (fn as (event: string, filename: string) => void)('rename', '01-Dec-2020_to_31-Dec-2020.csv');
      return ({ close: jest.fn() } as unknown) as FSWatcher;
    });

    mockdate.set('2020-01-01T00:00:00.000Z');
  });

  afterEach(() => {
    jest.resetAllMocks();
    mockdate.reset();
  });

  const mockUrls = (...urls: Array<string>) => {
    mocked(mocks.page.url).mockImplementation(() => urls.shift() ?? '');
  };

  describe('start', () => {
    it('should launch puppeteer with given options', async () => {
      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        puppeteerOpts: {
          defaultViewport: {
            height: 600,
            width: 800
          }
        }
      });

      await scraper.start();

      expect(mocked(puppeteer.launch)).toBeCalledWith({
        defaultViewport: {
          height: 600,
          width: 800
        }
      });
    });
  });

  describe('stop', () => {
    it('should close puppeteer', async () => {
      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        puppeteerOpts: {
          defaultViewport: {
            height: 600,
            width: 800
          }
        }
      });

      await scraper.start();
      await scraper.stop();

      expect(mocked(mocks.browser.close)).toBeCalled();
    });
  });

  describe('login', () => {
    it('should log in with given credentials', async () => {
      mockUrls('https://www.amazon.com/ap/signin');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await scraper.getItems().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith(
        'input[name=email]',
        'testuser@example.com'
      );

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=password]', 'test123');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberMe]');
    });

    it('should throw error if OTP code is required, but no OTP option is provided', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getItems().next()).rejects.toThrow(
        '2FA code requested, but neither otpSecret nor otpFn were provided'
      );
    });

    it('should login using generated OTP code if otpSecret is provided', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        otpSecret: 'GVYX4RKDFYXXMNDJ'
      });

      await scraper.getItems().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=otpCode]', '930899');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberDevice]');
    });

    it('should login using OTP code provided by otpFn', async () => {
      mockUrls('https://www.amazon.com/ap/signin', 'https://www.amazon.com/ap/mfa');

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123',
        otpFn: async () => '123456'
      });

      await scraper.getItems().next();

      expect(mocked(mocks.page.type)).toHaveBeenCalledWith('input[name=otpCode]', '123456');
      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('input[name=rememberDevice]');
    });

    it('should skip account fix-up page during login', async () => {
      mockUrls(
        'https://www.amazon.com/ap/signin',
        'https://www.amazon.com/ap/accountfixup',
        'https://www.amazon.com/ap/accountfixup'
      );

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await scraper.getItems().next();

      expect(mocked(mocks.page.click)).toHaveBeenCalledWith('a[id*=skip]');
    });

    it('should throw error if sign-out assertion fails', async () => {
      mockUrls('https://www.amazon.com/ap/signin');
      mocked(mocks.page.$).mockResolvedValue(null);

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getItems().next()).rejects.toThrow('No sign-out link detected');
    });

    it('should throw error on bad response', async () => {
      mockUrls('https://www.amazon.com/ap/signin');
      mocked(mocks.response.ok).mockReturnValue(false);

      const scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      await expect(scraper.getItems().next()).rejects.toThrow('Failed request');
    });
  });

  describe('getItems', () => {
    let scraper: AmazonScraper;

    beforeEach(() => {
      scraper = new AmazonScraper({
        username: 'testuser@example.com',
        password: 'test123'
      });

      mocked(createReadStream).mockImplementation(() =>
        jest.requireActual('fs').createReadStream(appRootPath.resolve('test-data/items.csv'))
      );
    });

    it('should return item for each row in report', async () => {
      const items: Array<OrderItem> = [];
      for await (const item of scraper.getItems()) {
        items.push(item);
      }

      expect(items).toHaveLength(3);
      expect(items).toMatchSnapshot();
    });

    it('should should save to correct directory', async () => {
      for await (const _ of scraper.getItems()) {
      }

      expect(mocked(mocks.cdpSession.send)).toBeCalledWith('Browser.setDownloadBehavior', {
        behavior: 'allow',
        downloadPath: '/tmp/amzscrKudNUt'
      });
    });

    it('should use provided dates to retrieve report', async () => {
      for await (const _ of scraper.getItems({
        startDate: new Date('2020-01-07T00:00:00.000Z'),
        endDate: new Date('2021-02-14T00:00:00.000Z')
      })) {
      }

      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-day-start', '6');
      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-month-start', '1');
      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-year-start', '2020');
      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-day-end', '13');
      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-month-end', '2');
      expect(mocked(mocks.page.select)).toHaveBeenCalledWith('#report-year-end', '2021');
    });

    it('should delete report after retrieval', async () => {
      const fakeElements = ([{ click: jest.fn() }] as unknown) as Array<ElementHandle<Element>>;
      mocked(mocks.page.$x).mockResolvedValue(fakeElements);

      for await (const _ of scraper.getItems()) {
      }

      expect(mocked(mocks.page.$x)).toHaveBeenCalledWith(
        '//input[@name="delete-report" and ancestor::tr[contains(., \'amzscr-5fb041e4-ad7a-41d4-879f-d1ec1919201a\')]]'
      );
      expect(fakeElements[0].click).toBeCalled();
    });

    it('should delete report on disk', async () => {
      const fakeElements = ([{ click: jest.fn() }] as unknown) as Array<ElementHandle<Element>>;
      mocked(mocks.page.$x).mockResolvedValue(fakeElements);

      for await (const _ of scraper.getItems()) {
      }

      expect(unlink).toBeCalledWith('/tmp/amzscrKudNUt/01-Dec-2020_to_31-Dec-2020.csv');
    });
  });
});
