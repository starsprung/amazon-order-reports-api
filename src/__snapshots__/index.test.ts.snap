// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmazonScraper getItems should return item for each row in report 1`] = `
Array [
  Object {
    "asinIsbn": "B06Y5XG33Q",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "USPS(9300120111405739275629)",
    "category": "HEALTH_PERSONAL_CARE",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 21.75,
    "itemSubtotalTax": 0,
    "itemTotal": 21.75,
    "listPricePerUnit": 0,
    "orderDate": "2020-01-03T00:00:00.000-08:00",
    "orderId": "123-7654321-1234567",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Visa - 1234",
    "purchasePricePerUnit": 21.75,
    "quantity": 1,
    "seller": "This Is A Seller",
    "shipmentDate": "2020-01-06T00:00:00.000-08:00",
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Some Product",
    "unspscCode": "38462846",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B07X9FXSPL",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "AMZN_US(TBA014759273649)",
    "category": "LIGHT_BULB",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 59.94,
    "itemSubtotalTax": 0,
    "itemTotal": 59.94,
    "listPricePerUnit": 24.99,
    "orderDate": "2020-04-20T00:00:00.000-07:00",
    "orderId": "321-9876543-3456789",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Visa - 4321",
    "purchasePricePerUnit": 19.98,
    "quantity": 3,
    "seller": "Amazon.com",
    "shipmentDate": "2020-04-21T00:00:00.000-07:00",
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Another Product",
    "unspscCode": "65927561",
    "website": "Amazon.com",
  },
  Object {
    "asinIsbn": "B0159OP2FF",
    "buyerName": "Test Man",
    "carrierNameTrackingNumber": "AMZN_US(TBA023749506528)",
    "category": "STANDALONE_STREAMING_MEDIA_PLAYER",
    "condition": "new",
    "currency": "USD",
    "itemSubtotal": 69.69,
    "itemSubtotalTax": 0,
    "itemTotal": 69.69,
    "listPricePerUnit": 69.69,
    "orderDate": "2020-06-09T00:00:00.000-07:00",
    "orderId": "987-3456789-7654321",
    "orderStatus": "Shipped",
    "orderingCustomerEmail": "email@example.com",
    "paymentInstrumentType": "Gift Certificate/Card and Visa - 5656",
    "purchasePricePerUnit": 69.69,
    "quantity": 1,
    "seller": "Amazon.com",
    "shipmentDate": "2020-06-10T00:00:00.000-07:00",
    "shippingAddressCity": "Washington",
    "shippingAddressName": "Test Man",
    "shippingAddressState": "DC",
    "shippingAddressStreet1": "1600 Pennsylvania Avenue NW",
    "shippingAddressZip": "20500",
    "title": "Yet Another Product",
    "unspscCode": "37462840",
    "website": "Amazon.com",
  },
]
`;
